source/robot_model/src/Model.cpp

```diff
@@ -260,9 +263,114 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+    }     
+  }
+  
+  return Psi;
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                              const state_representation::JointState& current_joint_state,
+                                                              std::string frame_name,
+                                                              const Model::InverseGeometryParameters& params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  } else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  // 1 second for the Newton-Raphson method
+  const std::chrono::nanoseconds dt((int)(1e9));
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+  state_representation::Jacobian J(this->get_robot_name(), this->robot_model_.nq, frame_name);
+  Eigen::MatrixXd J_b = J.data();
+  Eigen::MatrixXd JJt(6,6);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+  Eigen::VectorXd err(6);
+  state_representation::CartesianPose X(this->get_robot_name());
+  const state_representation::CartesianPose X_d(desired_cartesian_state);
+  
+  int i = 0;
+  while (i < params.max_number_of_iterations) {
+    X = this->forward_geometry(q, frame_id);
+    
+    err = ((X_d - X)/dt).data();
+
+    if(err.array().abs().maxCoeff() < params.tolerance) {
+      return q;
+    }
+
+    J = this->compute_jacobian(q);
+    
+    W_b = this->cwln_weighted_matrix(q, params.margin);
+    W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq) - W_b;
+    psi = params.gamma * this->cwln_repulsive_potential_field(q, params.margin);
+    J_b = J.data() * W_b;
+    
+    JJt.noalias() = J_b * J_b.transpose();
+    JJt.diagonal().array() += params.damp;
+    dq.set_velocities(W_c * psi + params.alpha * W_b * ( J_b.transpose() * JJt.ldlt().solve(err - J.data() * W_c * psi)));
+    
+    q = q + dq * dt;
+    q = this->clamp_in_range(q);
+
+    i++;
+  }
+  
+  throw (exceptions::IKDoesNotConverge(params.max_number_of_iterations, err.array().abs().maxCoeff()));
+}
+
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            std::string frame_name, const Model::InverseGeometryParameters& params) {
```

Comment:
`frame_name` can be passed by reference, `const std::string& frame_name`

Reply:
I do not agree because we change its value:
```
if (frame_name.empty()) {
    // get last frame if none specified
    frame_name = this->robot_model_.frames.back().name;
  } else {
    // throw error if specified frame does not exist
    if (!this->robot_model_.existFrame(frame_name)) {
      throw (exceptions::FrameNotFoundException(frame_name));
    }
  }
```

Reply:
Not in this function

Reply:
Thats the thing, if you use CLion, it will tell you all these little details

Reply:
again @patr3 unless there is a very valid reason all input variables should be made const reference. This is to prevent undesired side effect.

Reply:
You have convinced me. I install CLion +1